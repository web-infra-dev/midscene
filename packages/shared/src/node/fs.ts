import { existsSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { ifInBrowser, ifInWorker } from '../utils';

declare const __HTML_ELEMENT_SCRIPT__: string;

interface PkgInfo {
  name: string;
  version: string;
  dir: string;
}

const pkgCacheMap: Record<string, PkgInfo> = {};

export function getRunningPkgInfo(dir?: string): PkgInfo | null {
  if (ifInBrowser || ifInWorker) {
    return null;
  }
  const dirToCheck = dir || process.cwd();
  if (pkgCacheMap[dirToCheck]) {
    return pkgCacheMap[dirToCheck];
  }

  const pkgDir = findNearestPackageJson(dirToCheck);
  const pkgJsonFile = pkgDir ? join(pkgDir, 'package.json') : null;

  if (pkgDir && pkgJsonFile) {
    const { name, version } = JSON.parse(readFileSync(pkgJsonFile, 'utf-8'));
    pkgCacheMap[dirToCheck] = {
      name: name || 'midscene-unknown-package-name',
      version: version || '0.0.0',
      dir: pkgDir,
    };
    return pkgCacheMap[dirToCheck];
  }
  return {
    name: 'midscene-unknown-package-name',
    version: '0.0.0',
    dir: dirToCheck,
  };
}

/**
 * Find the nearest package.json file recursively
 * @param {string} dir - Home directory
 * @returns {string|null} - The most recent package.json file path or null
 */
export function findNearestPackageJson(dir: string): string | null {
  const packageJsonPath = join(dir, 'package.json');
  if (existsSync(packageJsonPath)) {
    return dir;
  }

  const parentDir = dirname(dir);

  // Return null if the root directory has been reached
  if (parentDir === dir) {
    return null;
  }

  return findNearestPackageJson(parentDir);
}

export function getElementInfosScriptContent() {
  const htmlElementScript = __HTML_ELEMENT_SCRIPT__;

  if (!htmlElementScript) {
    throw new Error('HTML_ELEMENT_SCRIPT inject failed.');
  }
  return htmlElementScript;
}

export async function getExtraReturnLogic(tree = false) {
  if (ifInBrowser || ifInWorker) {
    return null;
  }

  const elementInfosScriptContent = `${getElementInfosScriptContent()}midscene_element_inspector.setNodeHashCacheListOnWindow();`;

  if (tree) {
    return `${elementInfosScriptContent}midscene_element_inspector.webExtractNodeTree()`;
  }
  return `${elementInfosScriptContent}midscene_element_inspector.webExtractTextWithPosition()`;
}
