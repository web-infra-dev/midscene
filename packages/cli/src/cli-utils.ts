import { glob } from 'glob';
import 'dotenv/config';
import { statSync } from 'node:fs';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import { hideBin } from 'yargs/helpers';
import yargs from 'yargs/yargs';

declare const __VERSION__: string;

export const parseProcessArgs = async (): Promise<{
  path?: string;
  options: Record<string, any>;
}> => {
  const args = yargs(hideBin(process.argv))
    .usage(
      `Midscene.js helps you automate browser actions, assertions, and data extraction by AI. 
Homepage: https://midscenejs.com/
Github: https://github.com/web-infra-dev/midscene

Usage: $0 [options] <path-to-yaml-script-file-or-directory>`,
    )
    .options({
      headed: {
        type: 'boolean',
        default: false,
        description: 'Run the browser in headed mode to see the browser UI',
      },
      'keep-window': {
        type: 'boolean',
        default: false,
        description:
          'Keep the browser window open after the script finishes. This is useful when debugging, but will consume more resources',
      },
    })
    .version('version', 'Show version number', __VERSION__)
    .help()
    .wrap(yargs().terminalWidth());

  const argv = await args.argv;

  return {
    path: argv._[0] as string | undefined,
    options: argv,
  };
};

// match yml or yaml files
export async function matchYamlFiles(fileGlob: string) {
  if (existsSync(fileGlob) && statSync(fileGlob).isDirectory()) {
    fileGlob = join(fileGlob, '**/*.{yml,yaml}');
  }
  const files = await glob(fileGlob, {
    nodir: true,
    windowsPathsNoEscape: true,
    ignore: ['**/node_modules/**'],
  });
  return files
    .filter((file) => file.endsWith('.yml') || file.endsWith('.yaml'))
    .sort();
}
