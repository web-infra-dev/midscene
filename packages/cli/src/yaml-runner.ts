import { readFileSync } from 'node:fs';
import { basename, extname } from 'node:path';
import { ScriptPlayer, parseYamlScript } from '@midscene/web/yaml';
import { createServer } from 'http-server';
import {
  type MidsceneYamlFileContext,
  contextInfo,
  contextTaskListSummary,
  isTTY,
  singleTaskInfo,
  spinnerInterval,
} from './printer';
import { TTYWindowRenderer } from './tty-renderer';

import { assert } from 'node:console';
import type { FreeFn } from '@midscene/core';
import { puppeteerAgentForTarget } from '@midscene/web/puppeteer';

export const launchServer = async (
  dir: string,
): Promise<ReturnType<typeof createServer>> => {
  // https://github.com/http-party/http-server/blob/master/bin/http-server
  return new Promise((resolve, reject) => {
    const server = createServer({
      root: dir,
    });
    server.listen(0, '127.0.0.1', () => {
      resolve(server);
    });
  });
};

let ttyRenderer: TTYWindowRenderer | undefined;
export async function playYamlFiles(
  files: string[],
  options?: {
    headed?: boolean;
    keepWindow?: boolean;
  },
): Promise<boolean> {
  // prepare
  const fileContextList: MidsceneYamlFileContext[] = [];
  for (const file of files) {
    const script = parseYamlScript(readFileSync(file, 'utf-8'), file);
    const fileName = basename(file, extname(file));
    const preference = {
      headed: options?.headed,
      keepWindow: options?.keepWindow,
      testId: fileName,
    };
    const player = new ScriptPlayer(
      script,
      async (target) => {
        const freeFn: FreeFn[] = [];

        // launch local server if needed
        let localServer: Awaited<ReturnType<typeof launchServer>> | undefined;
        let urlToVisit: string | undefined;
        assert(typeof target.url === 'string', 'url is required');
        if (target.serve) {
          localServer = await launchServer(target.serve);
          const serverAddress = localServer.server.address();
          freeFn.push({
            name: 'local_server',
            fn: () => localServer?.server.close(),
          });
          if (target.url.startsWith('/')) {
            urlToVisit = `http://${serverAddress?.address}:${serverAddress?.port}${target.url}`;
          } else {
            urlToVisit = `http://${serverAddress?.address}:${serverAddress?.port}/${target.url}`;
          }
          target.url = urlToVisit;
        }

        const { agent, freeFn: newFreeFn } = await puppeteerAgentForTarget(
          target,
          preference,
        );
        freeFn.push(...newFreeFn);

        return { agent, freeFn };
      },
      (taskStatus) => {
        if (!isTTY) {
          const { nameText } = singleTaskInfo(taskStatus);
          // console.log(`${taskStatus.status} - ${nameText}`);
        }
      },
    );
    fileContextList.push({ file, player });
  }

  // play
  if (isTTY) {
    const summaryContents = () => {
      const summary: string[] = [''];
      for (const context of fileContextList) {
        summary.push(
          contextTaskListSummary(context.player.taskStatusList, context),
        );
      }
      summary.push('');
      return summary;
    };
    ttyRenderer = new TTYWindowRenderer({
      outputStream: process.stdout,
      errorStream: process.stderr,
      getWindow: summaryContents,
      interval: spinnerInterval,
    });

    ttyRenderer.start();
    for (const context of fileContextList) {
      await context.player.run();
    }
    ttyRenderer.stop();
  } else {
    for (const context of fileContextList) {
      const { mergedText } = contextInfo(context);
      console.log(mergedText);
      await context.player.run();
      console.log(
        contextTaskListSummary(context.player.taskStatusList, context),
      );
    }
  }

  const ifFail = fileContextList.some((task) => task.player.status === 'error');
  return !ifFail;
}
