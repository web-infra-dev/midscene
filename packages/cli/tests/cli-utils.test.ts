import { join } from 'node:path';
import { matchYamlFiles, parseArgsIntoYamlScript } from '@/cli-utils';
import { launchServer } from '@/yaml-runner';
import { describe, expect, test } from 'vitest';
const serverRoot = join(__dirname, 'server_root');

describe('cli utils', () => {
  test('match exact file', async () => {
    const files = await matchYamlFiles('./tests/midscene_scripts/foo.yml');
    expect(files).toMatchSnapshot();
  });

  let files: string[];
  test('match folder', async () => {
    files = await matchYamlFiles('./tests/midscene_scripts/');
    expect(files).toMatchSnapshot();
  });

  test('match folder 2', async () => {
    const files2 = await matchYamlFiles('./tests/midscene_scripts');
    expect(files2).toEqual(files);
  });

  test('match folder with star', async () => {
    const files2 = await matchYamlFiles('./tests/midscene_scripts/**');
    expect(files2).toEqual(files);
  });

  test('match files', async () => {
    const files3 = await matchYamlFiles('./tests/midscene_scripts/**/*.yml');
    expect(files3).toMatchSnapshot();
  });
});

describe('launch server', () => {
  test('launch server', async () => {
    const serverResult = await launchServer(serverRoot);
    expect(serverResult).toBeDefined();

    const serverAddress = serverResult.server.address();
    const staticServerUrl = `http://${serverAddress?.address}:${serverAddress?.port}`;

    const contents = await fetch(`${staticServerUrl}/index.html`);
    expect(contents.status).toBe(200);

    await serverResult.server.close();
  });
});
