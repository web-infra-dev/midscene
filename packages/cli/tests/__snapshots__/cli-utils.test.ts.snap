// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli utils > match exact file 1`] = `[]`;

exports[`cli utils > match files 1`] = `
[
  "tests/midscene_scripts/local/local.yml",
]
`;

exports[`cli utils > match folder 1`] = `
[
  "tests/midscene_scripts/local/local.yml",
  "tests/midscene_scripts/online/todomvc.yaml",
]
`;
